### DISCUSSION TITLE
Create branch directly from issue UI

### DISCUSSION THREAD

**Discussion by ajRiverav** [community/community#3441] [Projects and Issues] [answered]
# NOTE: UPVOTE THIS PROPOSAL IF YOU LIKE IT

![image](https://user-images.githubusercontent.com/22076301/109396401-bd254380-7907-11eb-91f7-75941b32e400.png) rather than liking the discussion along with the discussion's "👍". 

# PROPOSAL 
Write-up  by https://github.com/krnlde (originally at https://github.com/isaacs/github/issues/1125)

Coming from Jira [and @ajRiverav from Gitlab] [we] really miss the possibility to create a [feature|hotfix] branch that matches the current issue with its name and ID. I suggest to put this workflow in the sidebar of an issue where the Assignees, Labels and Milestones reside.

The following workflow would be nice:

Create an issue "My Example Issue" with id 42
See a "Create Branch for Issue"-button in the sidebar
Click on it and choose where to branch from (e.g. develop)
define the branch-type according to gitflow (feature/hotfix)
Create the branch whose name will be inferred from the issue's information feature/my-example-issue-42 or choose your own name
git fetch on your local machine and see the new branch
switch to it, make the necessary changes described in the issue
See the number of commits to this issue in the sidebar
Create a pull-request for the branch directly from the sidebar

**Comment by LifeIsStrange** [comment]
Just saying that this issue has 1533 likes! It is by far the most demanded feature request on github.com, will the company listen to its userbase or continue ignoring us?

**Comment by Shinigami92** [comment]
@ajRiverav At the top of your discussion description, could you write that folks should upvote the discussion along with the discussion's ":+1:" ? 
This will bring up the discussion higher if sorted by votes.

![image](https://user-images.githubusercontent.com/7195563/109395885-0d63cc00-792f-11eb-83c2-92486f482130.png)

**Reply by TPS** [reply]
Whoops, most of us aren't used to ↑ing (@ top left on mobile), so I didn't see it there.

**Reply by Shinigami92** [reply]
@TPS Yeah :confused: I visited this site extra on desktop so I can upvote it
Maybe we should tell GitHub :thinking:

**Reply by LifeIsStrange** [reply]
The redundancy between liking the issue and the issue description really is a bad UX pattern, come on github...

**Reply by ajRiverav** [reply]
that ambiguity is a different issue but noted and agreed ;)

**Reply by fonsp** [reply]
I opened a separate discussion about this: https://github.com/github/feedback/discussions/3474

**Reply by aron-hoogeveen** [reply]
I must say that I did not see the small hand symbol on that image and therefore it was not immediately clear where I should upvote. Maybe this can be changed to be more clear?

**Comment by JonathanLastrilla** [comment]
Come on github!

**Comment by ghost** [comment]
It's currently possible to do this with GitHub Apps, but a native integration would be nice.

**Reply by ajRiverav** [reply]
can you share references?

**Reply by ghost** [reply]
@ajRiverav I use the [create-issue-branch](https://github.com/robvanderleek/create-issue-branch) GitHub app which was specifically designed for the purpose proposed in the original issue. Works pretty well.

**Reply by ajRiverav** [reply]
I use that one, too.

**Reply by PabloAlexis611** [reply]
Thanks for the suggestion @GitGangGuy, I just tried it out and works nice!

**Comment by ysquaredd** [comment]
The original issue had 134 comments and 1547 thumbs up for this feature.  I hope that would be considered as part of this feedback.
![Feature_Request__Workflow_to_create_branch_from_issue_·_Issue__1125_·_isaacs_github](https://user-images.githubusercontent.com/12146822/110666996-19bb1500-81ca-11eb-9783-499020bc6a2e.png)

**Comment by nachoal** [comment]
Started to use GH for setting up a good project management workflow and this feature would be really useful. The original issue as commented from @YuraYefymenko has lots of demand since 2017, please add!

**Comment by rjk** [comment]
It sounds like this isn't very important ... till you get used to this workflow (e.g. in jira/bitbucket), then you can't live without it. 

It helps the project keep consistent branch names. It helps ensure developers follow the desired process by making it the easiest. And of course it simply makes branches for issues quick to create. 

Please Github, make my life 27% better and implement this feature.

**Comment by bwklein** [comment]
This already works really well in GitLab.  It creates the branch and the Merge Request (in Draft status) with one click.
https://docs.gitlab.com/ee/user/project/repository/web_editor.html#create-a-new-branch-from-an-issue

I just went to find out how to do this in GitHub and was disappointed to find out it didn't exist and found my way to this discussion by clicking through unresolved past issues.

**Comment by evi-liu** [comment]
Hey y'all. This feedback space is currently reserved for GitHub Discussions only. However, I've passed this feedback along to our Issues team so they can take a look.

**Reply by ajRiverav** [reply]
thx @evi-liu. How can we follow up in the future?

**Reply by jniclas** [reply]
Hey, has there been any progress on the feature in the last two months? 
Can't wait to see this on the Website :)

**Reply by ajRiverav** [reply]
@evi-liu

**Reply by TPS** [reply]
> Hey y'all. This feedback space is currently reserved for GitHub Discussions only. However, I've passed this feedback along to our Issues team so they can take a look.

This could now be moved to the new General category.

**Comment by robertwt7** [comment]
Come on github, lets do this 👍

**Comment by michellemerrill** [comment]
👋🏼 @krnlde @ajRiverav + everyone, thanks for the feedback here (and isaacs/github#1125)! Currently, this work is being prioritized internally. It will surface later on the [public roadmap](https://github.com/github/roadmap/projects/1). Keep watching this space!

**Reply by borsTiHD** [reply]
Awesome :D

**Reply by krnlde** [reply]
I must be dreaming. An awesome dream though

**Reply by krisjans** [reply]
Great news!
Such small conveniences and attention to detail distinguishes truly great product from good one.

**Comment by michellemerrill** [comment, marked as answer]
> 👋🏼 @krnlde @ajRiverav + everyone, thanks for the feedback here (and [isaacs/github#1125](https://github.com/isaacs/github/issues/1125))! Currently, this work is being prioritized internally. It will surface later on the [public roadmap](https://github.com/github/roadmap/projects/1). Keep watching this space!

📣 Update: This product enhancement published yesterday on our public roadmap! You can now track the issue [here](https://github.com/github/roadmap/issues/218).

**Reply by michellemerrill** [reply]
📣 hi all: We are now in public Beta: https://github.com/github/feedback/discussions/12290

**Reply by michellemerrill** [reply]
Also, locking this discussion, so we can centralize feedback about this feature at: https://github.com/github/feedback/discussions/12290

**Comment by glebsts** [comment]
So, it is currently in "Future" on roadmap, which means "not before q3 2022". Sounds like really complicated feature.

**Reply by TPS** [reply]
Per https://github.com/github/feedback/discussions/3441#discussioncomment-1926157, this now in beta.  Signup there if you wish.

**Reply by michellemerrill** [reply]
We are now in public beta: https://github.com/github/feedback/discussions/12290

**Comment by andreimarcut** [comment]
This feature could improve the productivity of my development team by using GitHub standard configuration. In turn, this will also improve the privacy of my private project because right now I have to share my private repository with 3rd party code only to achieve the desired productivity. Setting up a 3rd party solution is also a time-consuming task.

I hope it's getting prioritized and developed soon. 

I see only wins for everyone from every angle.

**Comment by dipree** [comment]
Hi 👋 I hope everyone is doing well! We are currently working on adding the functionality to allow creating a branch for an issue. If anyone is interested in trying it out early, please let me no with a comment here and I'll enable it for you.

**Reply by jimhill** [reply]
👋 Yes please

**Reply by Shinigami92** [reply]
:raising_hand:

**Reply by trevorhreed** [reply]
I'm down for trying it out.

**Reply by jniclas** [reply]
I'd be glad to!

**Reply by bb1950328** [reply]
Great! I also would like to try it out.

**Reply by kevinsuedmersen** [reply]
I'd be glad to check it out, too!

**Reply by dipree** [reply]
Enabled for all of the above. You will see a new section in issues called "Development" with a "Create a branch" button. If you are a member of the repo you can create branches linked to the issue for that particular repo. There are a few rough edges we still have to resolve. However, please don't hesitate to leave any feedback here.

For anyone else who's not in yet, I'll have a look once a day and add you.

**Reply by erickgarro** [reply]
I’m in too. Thanks!

**Reply by Shinigami92** [reply]
<img src="https://user-images.githubusercontent.com/7195563/148592388-b9da1933-a406-4792-966c-050fe8cad569.png" width="200px"/>

:+1: 

I think I will try this out tomorrow

**Reply by rjk** [reply]
yes please @dipree I'd like to try it out.

**Reply by robinpellegrims** [reply]
I'm interested as well :-)

**Reply by robvanderleek** [reply]
Interested also 🤚

**Reply by andreimarcut** [reply]
I am very happy to start using this feature and save time! Thank you. 🙏

**Reply by bb1950328** [reply]
Feedback from beta-testing:
1. Please add an option to do nothing after creating the branch. Currently the user can only choose between "clone the branch locally" and "Open in GitHub desktop". These are both things that I won't do because I will update the repo in my IDE.
2. The second dialog is a bit too narrow and the code block gets a horizontal scrollbar. I don't see a reason to make the dialog that narrow, it's only using about 25% of the screen width.
![image](https://user-images.githubusercontent.com/36961718/148656250-4b9558f2-b17f-4271-bfec-bb9853a4ba2e.png)

Other than that the feature looks well integrated and is very useful for me👍

**Reply by PhilippAuch** [reply]
Sounds great. I'd like to try it too :)

**Reply by dipree** [reply]
Everyone above ☝️ is in too.

@bb1950328 thank you, that's great feedback! Considering this with the next planned UI improvements.

**Reply by TheNerdGuyLulu** [reply]
🙋‍♂️ Here, please 🙌
@dipree

**Reply by ihor-sviziev** [reply]
I'd like to try it as well

**Reply by followsound** [reply]
Would love to try this too

**Reply by seangarner** [reply]
Yes please!

**Reply by zephimir** [reply]
I am interested please :)

**Reply by tattan** [reply]
I would like to try this please

**Reply by dipree** [reply]
Enabled above ☝️

**Reply by skyFabioCozz** [reply]
I would like to try this please

**Reply by ajRiverav** [reply]
I'd like to try this (original issue poster :))

**Reply by dipree** [reply]
Enabled above ☝️ 

Let me know if there's anything you think is missing. We did a few minor updates in the meantime and we are working on adding the option to create the branch in a different repository.

**Reply by LifeIsStrange** [reply]
I'd like to try this plz
@dipree

**Reply by BellettiR** [reply]
hi! i'm interested in trying it :) thanks!

**Reply by tanettrimas** [reply]
Hi, I would like to test this! :D

**Reply by dipree** [reply]
Enabled above ☝️ 

Thanks everyone for testing. I won't add anymore users from now on since we are close to releasing this feature.

**Reply by yash-lysn** [reply]
Hi @dipree thanks for the updates!  Any ETA on when this will be live? Very keen to use this at my work :)

I can see according to roadmap it's (first issue in future milestone) bit too far away to be released is that accurate?

<img width="1382" alt="Screen Shot 2022-01-21 at 9 31 03 am" src="https://user-images.githubusercontent.com/92718748/150432567-2ac1d336-b0c3-4114-a764-9f6ee65c5e25.png">

**Reply by pietrovismara** [reply]
Hi, I would like to test this thank you!

**Reply by jimrazmus** [reply]
I am interested in trying this new feature.

**Reply by ajRiverav** [reply]
> Enabled above ☝️ 
> 
> 
> 
> Thanks everyone for testing. I won't add anymore users from now on since we are close to releasing this feature.

hey, before you reply to this thread, read this 🙃

**Reply by Shinigami92** [reply]
sryly :rofl: 

![image](https://user-images.githubusercontent.com/7195563/151506309-a1288c5a-c6c3-4f08-abce-5027a5ff59f8.png)

Is it possible to lock discussion-sub-threads? Or just us **BOLD TEXT**

**Reply by michael-hollander-philly-DAO** [reply]
I'd like to test too!

**Reply by ajRiverav** [reply]
this is becoming a super funny thread 😂

**Reply by michael-hollander-philly-DAO** [reply]
> this is becoming a super funny thread 😂

Sorry!

**Reply by DamianMehers** [reply]
I too am feeling the pain - I check the issue page for a "create branch" whenever I am on it ... each time to be disappointed.  Such a standard feature on other systems such as Azure Devops.  I am always forgetting to "#Close <issue number" when creating a pull request, because I'm used to the Issue->Branch->Pull Request flow.

@dipree  Please consider adding new beta testers if the release is not imminent.

**Reply by pietrovismara** [reply]
@DamianMehers I realized now that 21 hours ago the related issue has moved from "planning" to " Q1 2022 – Jan-Mar". The "in design" label was also replaced by the "beta" one. It looks like it's only a matter of weeks which sounds perfectly reasonable to me.

**Comment by ea0723** [comment]
Nice work, Dan! While the PM position didn’t work out for me, I really enjoyed the conversation I had with you and Jason -  it’s so cool to see this solution coming to fruition! 😊

Hope you had a great holiday break and happy 2022!

Cheers,

Elena

From: Daniel Adams ***@***.***>
Date: Friday, January 7, 2022 at 11:46 AM
To: github/feedback ***@***.***>
Cc: Elena Newton ***@***.***>, Manual ***@***.***>
Subject: Re: [github/feedback] Create branch directly from issue UI (Discussion #3441)

Enabled for all of the above. You will see a new section in issues called "Development" with a "Create a branch" button. If you are a member of the repo can create branches linked to the issue for that particular repo. There are a few rough edges we still have to resolve. However, please don't hesitate to leave any feedback here.

For anyone else who's not in yet, I'll have a look once a day and add you.

**Reply by jimhill** [reply]
This is gonna be a great time saver for me so thanks. Just dropping a message to say, that the integration for opening in GitHub Desktop doesn't seem to work. When I choose (from Safari) to open the new branch in GitHub Desktop, the app opens but the current branch doesn't change.

**Reply by dipree** [reply]
Likewise @ea0723, hope you are doing well! 😊

**Reply by Bestra** [reply]
> the app opens but the current branch doesn't change.

@jimhill thanks for speaking up! We have a [PR open to fix that](https://github.com/github/special-projects/issues/628#issuecomment-961417807) at the moment.

**Reply by TheNerdGuyLulu** [reply]
The modal opens when in full detail, but the button doesn't seem to work inside the project board
![Screenshot 2022-01-14 at 18 39 59](https://user-images.githubusercontent.com/20543676/149567991-f787743e-7964-4ef1-b715-f05e8b85ee8a.png)
.

**Reply by Bestra** [reply]
@TheNerdGuyLulu thank you for finding that! I'll file an internal issue and we'll look into it.

**Reply by Bestra** [reply]
I have a PR open for the fix, it should be deployed by Monday.

**Comment by ajRiverav** [comment]
@dipree:  Tried the beta. The flow feels natural and the branch creation is intuitive + does work :) . So far, it solves the issue I posted originally, which was about creating a branch from the browser. 

cc: @Bestra 

## Issue 1
- After creating the branch, I observed 
<img width="395" alt="image" src="https://user-images.githubusercontent.com/22076301/149840471-c8c13aa0-cac8-4486-8be2-a3fa417e240f.png">

However, there was no PR created (at least it was not listed in the corresponding tab). 

See 
<img width="346" alt="image" src="https://user-images.githubusercontent.com/22076301/149840551-f29ca251-75a1-463e-8fa5-fb5b7d448d1c.png">

## Issue 2
The "copy" button did not seem to work on latest (as of today) Safari. 

<img width="528" alt="image" src="https://user-images.githubusercontent.com/22076301/149840687-df15b6ac-34a5-4b0b-b743-79018676cf9c.png">

It only copies the branch name, which may be sufficient, I guess. 


## Observation

It's very nice that one can see all the "associated" branches created -> 
<img width="397" alt="image" src="https://user-images.githubusercontent.com/22076301/149840777-e8513415-fc02-4145-b31a-592d2e789883.png">

**Reply by dipree** [reply]
@ajRiverav great feedback, thank you.
1. We have intentionally decided not to create a PR alongside the branch because it's the same state as the source branch and of no initial value. However, the user can create one anytime and in the Development section the branch will get replaced with the PR automatically.
2. Nice catch. We'll fix that.

**Reply by ajRiverav** [reply]
@dipree, just FYI: I noticed this was shown on a repo for which I did not have permissions. Clicking on the link does nothing, as expected. 

I mean, it's feels strange to see this as an option on a repo I can't work on. 

e.g. 
<img width="1517" alt="image" src="https://user-images.githubusercontent.com/22076301/150008725-62cc79cf-7d31-4c6e-87dc-9ebf6d3524eb.png">

**Reply by Bestra** [reply]
@ajRiverav thanks! We show this functionality in a few places and I'm pretty sure this is one that we didn't catch! We'll take care of it shortly.
